src/api/service.rs

```diff
@@ -1,82 +1,918 @@
+
+        Ok(Ipv4 {
+            method,
+            address,
+            netmask,
+            gateway,
+        })
+    }
+}
+
+impl FromProperties for Ipv6 {
+    fn from_properties(properties: &RefArgMap, prop_name: &'static str) -> Result<Self, PropertyError> {
+        let mut i = get_property_argiter(properties, prop_name)?;
+        let mut m: HashMap<String, String> = HashMap::new();
+        while let Some(key) = i.next().and_then(|k| k.as_str()) {
+            if let Some(val) = i.next().and_then(|v| v.as_str()) {
+                let _ = m.insert(key.to_string(), val.to_string());
+            }
+        }
+
+        let method = if let Some(method) = m.get(Ipv6Kind::Method.into()) {
+            Some(method.parse::<Ipv6Method>()?)
+        } else { None };
+        let address = m.get(Ipv6Kind::Address.into()).cloned();
+        let prefix_length = if let Some(val) = m.get(Ipv6Kind::PrefixLength.into()) {
+            Some(
+                val.parse::<u8>()
+                    .map_err(|e| {
+                        PropertyError::Cast(Cow::Owned(val.clone()))
+                    })
+                    ?)
+        } else { None };
+        let gateway = m.get(Ipv6Kind::Gateway.into()).cloned();
+        let privacy = if let Some(privacy) = m.get(Ipv6Kind::Privacy.into()) {
+            Some(privacy.parse::<Ipv6Privacy>()?)
+        } else { None };
+
+        Ok(Ipv6 {
+            method,
+            address,
+            prefix_length,
+            gateway,
+            privacy,
+        })
+    }
+}
+
+impl FromProperties for Proxy {
+    fn from_properties(properties: &RefArgMap, prop_name: &'static str) -> Result<Self, PropertyError> {
+        let mut i = get_property_argiter(properties, prop_name)?;
+        let mut m: HashMap<String, &RefArg> = HashMap::new();
+        while let Some(key) = i.next().and_then(|k| k.as_str()) {
+            if let Some(val) = i.next() {
+                let _ = m.insert(key.to_string(), val);
+            }
+        }
+
+        let method = m.get(ProxyKind::Method.into())
+            .and_then(|refarg| refarg.as_str())
+            .and_then(|s| ProxyMethod::from_str(s).ok());
+
+        let url = m.get(ProxyKind::Url.into())
+            .and_then(|refarg| refarg.as_str())
+            .map(String::from);
+
+        let servers = m.get(ProxyKind::Servers.into())
+            .and_then(|refarg| cast::<Vec<String>>(&refarg.box_clone()).cloned());
+
+        let excludes = m.get(ProxyKind::Excludes.into())
+            .and_then(|refarg| cast::<Vec<String>>(&refarg.box_clone()).cloned());
+
+        Ok(Proxy {
+            method,
+            url,
+            servers,
+            excludes,
+        })
+    }
+}
+
+impl FromProperties for Provider {
+    fn from_properties(properties: &RefArgMap, prop_name: &'static str) -> Result<Self, PropertyError> {
+        let mut i = get_property_argiter(properties, prop_name)?;
+        let mut m: HashMap<String, String> = HashMap::new();
```

Comment:
I got away with pre-building the Map for most Dict-based properties since all fields are strings, and with Ethernet's precarious `EthernetKind` match approach because all fields are `Options`. Neither of these solutions are super ideal, but they work and future optimization shouldn't break compatibility 🤞. But the least I can do here is just store the `&str` to save on some unnecessary allocation.